actionpack/lib/action_controller/metal/rate_limiting.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module ActionController # :nodoc:
+  module RateLimiting
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      # Applies a rate limit to all actions or those specified by the normal <tt>before_action</tt> filters with <tt>only:</tt> and <tt>except:</tt>.
+      #
+      # The maximum number of requests allowed is specified <tt>to:</tt> and constrained to the window of time given by <tt>within:</tt>.
+      #
+      # Rate limits are by default unique to the ip address making the request, but you can provide your own identity function by passing a callable
+      # in the <tt>by:</tt> parameter. It's evaluated within the context of the controller processing the request.
+      #
+      # Requests that exceed the rate limit are refused with a <tt>429 Too Many Requests</tt> response. You can specialize this by passing a callable
+      # in the <tt>with:</tt> parameter. It's evaluated within the context of the controller processing the request.
+      #
+      # Examples:
+      #
+      #   class SessionsController < ApplicationController
+      #     rate_limit to: 10, within: 3.minutes, only: :create
+      #   end
+      #
+      #   class SignupsController < ApplicationController
+      #     rate_limit to: 1000, within: 10.seconds,
+      #       by: -> { request.domain }, with: -> { redirect_to busy_controller_url, alert: "Too many signups on domain!" }, only: :new
+      #   end
+      #
+      # Note: Rate limiting relies on the application having an accessible Redis server and on Kredis 1.7.0+ being available in the bundle.
+      # This uses the Kredis limiter type underneath, which is failsafe, so in case Redis is inaccessible, the rate limit will not refuse action execution.
+      def rate_limit(to:, within:, by: -> { request.remote_ip }, with: -> { head :too_many_requests }, **options)
```

Comment:
Love the api and explicitness of the error for required gem dependencies

Why this approach over creating an engine that is bootstrapped with a generator?

Reply:
I'm not sure I follow? An engine to provide rate limiting? Don't think we need to maintain another package just to provide one additional method in the public API.